lib/hooks/userspace/hook-koa.js

```diff
@@ -0,0 +1,130 @@
+var semver = require('semver');
+var constants = require('../../constants.js');
+var agent;
+
+var SUPPORTED_VERSIONS = '1.x';
+
+function useWrap(use) {
+  return function useTrace() {
+    if (!this._google_trace_patched) {
+      this._google_trace_patched = true;
+      this.use(middleware);
+    }
+    return use.apply(this, arguments);
+  };
+}
+
+function *middleware(next) {
+  var namespace = cls.getNamespace();
+  if (!namespace) {
+    agent.logger.info('Koa: no namespace found, ignoring request');
+    return;
+  }
+  if (!agent.shouldTrace(this.req.url)) {
+    return;
+  }
+  var req = this.req;
+  var res = this.res;
+
+  namespace.bindEmitter(req);
+  namespace.bindEmitter(res);
+
+  var originalEnd = res.end;
+
+  namespace.run(function() {
+    var rootContext = startRootSpanForRequest(req);
+
+    // wrap end
+    res.end = function(chunk, encoding) {
+      res.end = originalEnd;
+      var returned = res.end(chunk, encoding);
+
+      endRootSpanForRequest(rootContext, req, res);
+      return returned;
+    };
+    namespace.bind(next);
+  });
+  yield next;
+}
+
+/**
+ * Creates and sets up a new root span for the given request.
+ * @param {Object} req The request being processed.
+ * @returns {!SpanData} The new initialized trace span data instance.
+ */
+function startRootSpanForRequest(req) {
+  var result = agent.parseContextFromHeader(
+    req.headers[constants.TRACE_CONTEXT_HEADER_NAME]) || {};
+
+  var traceId = result.traceId;
+  var parentSpanId = result.spanId;
+  var url = (req.headers['X-Forwarded-Proto'] || 'http') +
+    '://' + req.headers.host + req.url;
+
+  // we use the path part of the url as the span name and add the full
+  // url as a label
+  var rootContext = agent.createRootSpanData(req.url, traceId,
+    parentSpanId);
+  rootContext.addLabel(TraceLabels.HTTP_METHOD_LABEL_KEY, req.method);
+  rootContext.addLabel(TraceLabels.HTTP_URL_LABEL_KEY, url);
+  rootContext.addLabel(TraceLabels.HTTP_SOURCE_IP, req.connection.remoteAddress);
+  return rootContext;
+}
+
+
+/**
+ * Ends the root span for the given request.
+ * @param {!SpanData} rootContext The span to close out.
+ * @param {Object} req The request being processed.
+ * @param {Object} res The response being processed.
+ */
+function endRootSpanForRequest(rootContext, req, res) {
+  if (req.route && req.route.path) {
+    rootContext.addLabel(
+      'koa/request.route.path', req.route.path);
+  }
+  rootContext.addLabel(
+      TraceLabels.HTTP_RESPONSE_CODE_LABEL_KEY, res.statusCode);
+  rootContext.close();
+}
+
+module.exports = function(version_, agent_) {
+  if (!semver.satisfies(version_, SUPPORTED_VERSIONS)) {
+    agent_.logger.info('Koa: unsupported version ' + version_ + ' loaded');
+    return {};
+  }
+  return {
+    // An empty relative path here matches the root module being loaded.
+    '': {
+      patch: function(koa) {
+        agent = agent_;
+        shimmer.wrap(koa.prototype, 'use', useWrap);
```

Comment:
To my knowledge, `shimmer` does not provide a way to patch a module itself but only properties of it. This could be done custom w/o shimmer if it is really required but it involves messing with the node module cache (or at least that is how I have investigated it in the past).